CacheDetailActivity: improve color detection

okainov opened the pull request:
Use W3 recommendations for determining contrast to decide whether to use neutral gray color.

Reference
https://www.w3.org/TR/2008/REC-WCAG20-20081211/#visual-audio-contrast-contrast

Fixes #7153

Lineflyer commented:
ok to test

okainov commented on tests/src-android/cgeo/geocaching/utils/ColorUtilsTest.java:
```diff
@@ -0,0 +1,26 @@
+
+import junit.framework.TestCase;
+
+import static org.assertj.core.api.Java6Assertions.assertThat;
```
I just used Androd Studio "Reformat code \ Optimize Import"... Changed the settings now, done

okainov commented on main/src/cgeo/geocaching/utils/ColorUtils.java:
```diff
@@ -0,0 +1,45 @@
+     * @param rgb2 second {@link Color} int
+     * @return contrast in range from 1 to 21
+     */
+    public static double getContrast(final int rgb1, final int rgb2) {
```
done

okainov commented on main/src/cgeo/geocaching/CacheDetailActivity.java:
```diff
@@ -1,5 +1,82 @@
+import java.util.regex.Pattern;
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
```
I hope, done

UniQP replied:
The `org.apache.commons` imports are still moved to a undesired position.

okainov replied:
done

okainov commented on main/src/cgeo/geocaching/CacheDetailActivity.java:
```diff
@@ -175,21 +177,14 @@
+
+    private static final Pattern[] COLOR_PATTERNS = {
+            // <span style="color: #0000ff;">
+            Pattern.compile("(?<!background-)color: ?\"?(#[0-9a-f]{6}|\\w+)", Pattern.CASE_INSENSITIVE),
```
Quote sign is missing in the suggestion, but for whitespace yes, replaced.

UniQP replied:
Is a quote sign allowed at this place?

okainov replied:
Probably missed this point, thanks.

okainov commented on main/src/cgeo/geocaching/CacheDetailActivity.java:
```diff
@@ -175,21 +177,14 @@
+            // <span style="color: #0000ff;">
+            Pattern.compile("(?<!background-)color: ?\"?(#[0-9a-f]{6}|\\w+)", Pattern.CASE_INSENSITIVE),
+            // <font color="red">
+            Pattern.compile("(?<!bg)color= ?\"?(#[0-9a-f]{6}|\\w+)", Pattern.CASE_INSENSITIVE)};
```
done

okainov commented on main/src/cgeo/geocaching/CacheDetailActivity.java:
```diff
@@ -1782,29 +1777,28 @@ private void addWarning(final UnknownTagsHandler unknownTagsHandler, final Spann
+            final MatcherWrapper matcher = new MatcherWrapper(colorPattern, descriptionString);
+            while (matcher.find()) {
+                try {
+                    final int textColor = Color.parseColor(matcher.group(0));
```
done

okainov commented:
By manually going thru my lists, found a few examples:

https://coord.info/GC1Y4B8 - "usually have an oval" is written in black
https://www.geocaching.com/geocache/GCM03P_united-colors?guid=89cec7d3-7c3d-4096-afe6-45eb4a0615d7 this looks ugly with explicit black font.

However I'm not a big expert in COM caches, so maybe @SammysHP can share some of hundreds caches with colors he had seen.

UniQP commented on main/src/cgeo/geocaching/CacheDetailActivity.java:
```diff
@@ -1734,7 +1729,7 @@ private void loadDescription(final String descriptionString, final IndexOutOfBou
                     fixRelativeLinks(descriptionView);
                     descriptionView.setMovementMethod(AnchorAwareLinkMovementMethod.getInstance());
-                    fixTextColor(descriptionString, descriptionView);
+                    fixTextColor(Html.toHtml(description), descriptionView);
```
Can you add a comment that explains why not to use `descriptionString` (similar to the explanation in the commit message).

okainov replied:
Just found out that there is already proper span parsing in fixRelativeUrl method, so I thew away all HTML parsing with regexps and replaced it to neat span parsing. Sorry for several iterations

Lineflyer commented:
@UniQP If you are fine with the code itself, feel free to merge to test it in nightlies. I will scan certain listings and see if the behaviour improved.

okainov commented:
@UniQP do you want me to squash commits at some point to have more neat history? I saw sometimes you merge it without squashing (usually GitHub allows you to merge PR and squash automatically), so it's not really clear to have 15 commits like "fix code style"...

UniQP commented:
@okainov I tested the code with GCM03P and it seems that setting the background color doesn't work.

> @UniQP do you want me to squash commits at some point to have more neat history? I saw sometimes you merge it without squashing […]

I will merge without squashing only if the commit history is clean (e.g. contains no fixup commits). You can decide yourself if you want to create a clean commit history in the end or if squashing all changes into a single commit is the right thing to do.

okainov commented:
>I tested the code with GCM03P and it seems that setting the background color doesn't work.

Good catch, fixed it now, postprocessing was done after setting the text.

I squashed and rebased all the commits since they all were related to this PR.

UniQP commented:
I continued testing and noticed that CSS colors aren't working anymore. The reason is that `Spannable` does not support `<style>` tags and CSS colors.

okainov commented:
@UniQP could you please provide cache where you observe this?

UniQP commented:
Strange: I tested GC54REW where I didn't see an effect. On the other hand, it works for GC2PAEP.

okainov commented:
>Strange: I tested GC54REW where I didn't see an effect. On the other hand, it works for GC2PAEP.

GC54REW has `<div style="background-color:#8597BB;color:#000000` in its descriptionString.
GC2PAEP has usual spans `<span style="color:#000000;">` in its descriptionString.

Html.toHtml(Html.fromHtml(string)) removes all the divs because it's not supported. It doesn't depend on this PR.

Original implementation behavior:
GC54REW - triggers gray background.  However it's useless, because, as noticed before, color in divs is not supported by Html.fromHtml. So even if one explicitly disables `fixColorText`, content will be perfectly readable.
GC2PAEP  - no fix, only black text, not readable. Reason why it wan't triggering on "span style" - lack of space in regexp betweeen ":" and "#00"

This PR implementation behavior:
GC54REW - no fix, since there is no color problems.
GC2PAEP - triggers gray background for part of description which is inside  `<span style="color:#000000;">`

UniQP commented:
Thank you for the investigation. I agree that it's a good thing to support exactly the HTML tags that are supported by the TextView.

okainov commented:
@rsudev @UniQP @SammysHP @Lineflyer  Just a friendly reminder, any updates?
